rapid_pro_tools/rapid_pro_client.py

```diff
@@ -0,0 +1,120 @@
+        raw_contacts = self.rapid_pro.get_contacts().all(retry_on_rate_exceed=True)
+        assert len(set(c.uuid for c in raw_contacts)) == len(raw_contacts), "Non-unique contact UUID in RapidPro"
+        print(f"Fetched {len(raw_contacts)} contacts")
+        return raw_contacts
+
+    def get_traced_runs_for_flow_id(self, user, flow_id, phone_uuids,
+                                    range_start_inclusive=None, range_end_exclusive=None,
+                                    test_contacts=None):
+        if test_contacts is None:
+            test_contacts = []
+
+        raw_runs = self.get_raw_runs_for_flow_id(flow_id, range_start_inclusive, range_end_exclusive)
+        raw_contacts = self.get_raw_contacts()
+        
+        contacts_lut = {c.uuid: c for c in raw_contacts}
+
+        traced_runs = []
+        for run in raw_runs:
+            if run.contact.uuid not in contacts_lut:
+                # Sometimes contact uuids which appear in `runs` do not appear in `contact_runs`.
+                # I have only observed this happen for contacts which were created very recently.
+                # This test skips the run in this case; it should be included next time this script is executed.
+                print(f"Warning: Run found with Rapid Pro Contact UUID '{run.contact.uuid}', "
+                      f"but this id is not present in the downloaded contacts")
+                continue
+
+            contact_urns = contacts_lut[run.contact.uuid].urns
+            if len(contact_urns) == 0:
+                print(f"Warning: Ignoring contact with no urn. URNs: {contact_urns} "
+                      f"(Rapid Pro Contact UUID: {run.contact.uuid})")
+                continue
+
+            run_dict = {
+                "avf_phone_id": phone_uuids.add_phone(contact_urns[0]),
+                f"run_id - {run.flow.name}": run.id
+            }
+
+            for category, response in run.values.items():
+                run_dict[category.title() + " (Category) - " + run.flow.name] = response.category
+                run_dict[category.title() + " (Value) - " + run.flow.name] = response.value
+                # Convert from "input" to "text" here to match terminology in Rapid Pro's Excel exports.
+                run_dict[category.title() + " (Text) - " + run.flow.name] = response.input
+                run_dict[category.title() + " (Name) - " + run.flow.name] = response.name
+                run_dict[category.title() + " (Time) - " + run.flow.name] = response.time.isoformat()
+                run_dict[category.title() + " (Run ID) - " + run.flow.name] = run.id
+
+            if run.contact.uuid in test_contacts:
+                run_dict["test_run"] = True
+            else:
+                assert len(contact_urns) == 1, \
+                    f"A non-test contact has multiple URNs (Rapid Pro Contact UUID: {run.contact.uuid})"
+
+            run_dict[f"created_on - {run.flow.name}"] = run.created_on.isoformat()
```

Comment:
These might benefit form being run_created_on, ....

However it's a bit marginal so if this causes massive downstream upheaval don't worry about it.

Reply:
changed